### ISSUE TITLE
FR: Free zoom

### ISSUE THREAD

**Issue by Frenzie** [koreader/koreader#5524]
* KOReader version: 2019.10
* Device: all

#### Issue
While the various zoom to fit X options are fantastic and cover just about any document I normally come across, messing about with cropping just to zoom to a section of the page is arduous at best. Page flipping mode features a slightly more free zoom & pan but it doesn't always work out.

Here's how it used to work in KPV (from [here](https://www.mobileread.com/forums/showthread.php?p=3906971#post3906971)), which is fairly straightforward:

> In KPV (Kindle DX) I can zoom out/in using ALT + <> (prev/next page) buttons, by predefined zoom step e.g. 16 %. (using SHIFT + <> buttons for zoom step decrease/increase)
>
>I can pan the active view using joypad stick (moving it up/down/left/right), also by predifined panning step e.g. 60 pixels.
>
>SHIFT + left/right would decrease/increase predifined x panning step.
>
>SHIFT + up/down would decrease/increase y panning step.

**Comment by Frenzie** [comment]
Copying from https://github.com/koreader/koreader-base/issues/1148#issuecomment-661894276

* The zoom to box feature is great *as a starting point* and it's neat most of the time but it doesn't always work out. You want much freer zoom.
  Some details on things we might want in free zoom here: https://github.com/koreader/koreader/issues/5524
* Ideally all this would somehow be available in the main reader mode easily without having to trigger a mostly hidden special mode.
  I'm not sure page flipping mode still has much purpose with the greatly improved skim widget.

**Comment by Frenzie** [comment]
Copying from https://github.com/koreader/koreader/pull/6408#issuecomment-661877691

We actually already have a zoom implemented in page flipping mode, although I just noticed something about the double tap there might be broken. https://github.com/koreader/koreader/pull/6322 is a potential suspect but it'd need to be investigated. Also see https://github.com/koreader/koreader/issues/5524

If you run KOReader on a multitouch device then you can also enter page flipping mode and zoom in/out with pinch gestures. As you'll see it's more of a guided zoom, not truly free.

https://github.com/koreader/koreader/blob/86db43a0527e94f1ebe2a124dae92f8fd3f70e6e/frontend/apps/reader/modules/readerzooming.lua#L315-L331

The main obstacle is that I haven't figured out if there's a good way to combine free scroll & zoom with the ease of use of zoom to fit content/page width, and also some undefined behavior outside of the page proper.

But most or perhaps all of the puzzle pieces are already in place. The guided zoom to text columns is a great concept — but only when it works nicely.

**Comment by poire-z** [comment]
Is this still another request/mode that what's been made available with #6885 and/or panel zoom?

**Comment by Galunid** [comment]
It's still another request. It's about zooming like you would on android, so not covered by #6885

**Comment by Frenzie** [comment]
Indeed, this is about having something like a +/- button and/or two finger pinch/spread gestures.

**Comment by hideandseekstheorder** [comment]
There any update in this 2023

**Comment by gd4c** [comment]
I am reading a PDF with somewhat-detailed figures that need zoom. I'll take any implementation. Even if the entire page is converted to image on long tapping an image and sent to KOReader's image viewer!

**Comment by NemesisHubris** [comment]
Just adding my support for this feature request! It’s still very relevant in 2025.

Kindle’s built-in PDF reader already allows pinch-to-zoom, panning around freely, and shows where you are on the page with a viewport indicator. That makes it much easier to read textbooks and detailed PDFs. In KOReader, I often try to read textbook PDFs where the text is too small, but current zoom options are limited, and I lose context when trying to enlarge the content. Tell me this weird situation where I actually use send to kindle to read pdfs on a jailbroken kindle with koreader installed.

What I’d love to see is a zoom system similar to what KOReader already provides for images: allow pinch-to-zoom, then pan around freely, with an optional small overlay/minimap showing the current viewport. A quick way to reset back to fit-to-page would also help. 

I’ve tried a few workarounds: panel zoom (useful only in comics), fit-width / zoom-to-content / margin cropping (helps but not enough for textbooks), and even gesture shortcuts to change zoom (not reliable, not intuitive, and doesn’t allow smooth free-moving).

Device: Kindle Paperwhite 12th Gen
KOReader version: newest build (as of today)

**Comment by shuvashish76** [comment]
Newspapers/Magazines/Academic PDF books are most problematic.
e.g. Red boxes are my "Zoom factor" and I want to read Red boxed ones only but currently I've to go through the _unnecessary full Blue lined boxes step_ to reach the destination red box. And if you try to change the "Direction" anywhere then the page jumps/resets to specific corner...you've to start all over again it's very messy.

<details><summary>Details</summary>
<p>

<img width="1200" height="2090" alt="Image" src="https://github.com/user-attachments/assets/8815676b-5e39-480d-bae0-5d3ec01abae8" />

</p>
</details>

Also imagine reading the "Fill Disclosure" section with extra zoom. 1st in zoom mode you've to read left-to right then where the line ends you've to go next few lines below that zoom but as mentioned you can't change direction without resetting the whole page.
